actors/evm/src/interpreter/system.rs

```diff
@@ -385,6 +410,89 @@ impl<'r, RT: Runtime> System<'r, RT> {
+                origin: origin_id,
+                nonce: self.rt.message().nonce(),
+            }),
+            Err(_) => None, // Handle the error case by returning None
+        }
+    }
+
+    /// Get value of a transient storage key.
+    pub fn get_transient_storage(&mut self, key: U256) -> Result<U256, ActorError> {
+        if self.transient_data_lifespan == self.get_current_transient_data_lifespan() {
+            // Lifespan matches, retrieve value
+            Ok(self
+                .transient_slots
+                .get(&key)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to access transient storage slot")?
+                .cloned()
+                .unwrap_or_default())
+        } else {
+            // Lifespan mismatch, return default value
+            Ok(U256::zero())
+        }
+    }
+
+
+    /// Set value of a transient storage key.
+    pub fn set_transient_storage(&mut self, key: U256, value: U256) -> Result<(), ActorError> {
+        let current_lifespan = self.get_current_transient_data_lifespan();
+
+        if self.transient_data_lifespan == current_lifespan {
+            // Lifespan matches, proceed to set value
+            let changed = if value.is_zero() {
+                self.transient_slots
+                    .delete(&key)
+                    .map(|v| v.is_some())
+                    .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to clear transient storage slot")?
+            } else {
+                self.transient_slots
+                    .set(key, value)
+                    .map(|v| v != Some(value))
+                    .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to update transient storage slot")?
+            };
+
+            if changed {
+                self.saved_state_root = None; // Mark state as dirty
+            }
+        } else {
+            // Lifespan mismatch
+            if value.is_zero() {
+                // If setting to default value, skip storage reset
+                return Ok(());
+            }
+
+            // Reset transient storage and update lifespan
+            self.reset_transient_storage(current_lifespan)?;
+            self.transient_slots
+                .set(key, value)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to set transient storage after reset")?;
+
+        }
+
+        Ok(())
+    }
+
+    /// Reset transient storage and update lifespan.
+    pub fn reset_transient_storage(&mut self, new_lifespan: Option<TransientDataLifespan>) -> Result<(), ActorError> {
+        // Update lifespan
+        self.transient_data_lifespan = new_lifespan;
+
+        // Reinitialize the transient_slots with a fresh KAMT
+        //let transient_store = self.rt.store().clone();
+        //self.transient_slots = StateKamt::new_with_config(transient_store, KAMT_CONFIG.clone());
+        // TODO XXX reinitialize does not currently work due to blockstore reference issues
```

Comment:
I'd be interested to know what these errors are and how you end up diagnosing the problem, for my own educational purposes because your commented code looks like it should work.

Reply:
It's because `self.rt.store()` doesn't implement clone, while `&...` does, so rust is helpfully creating a reference for you.

Note: the `new` function explicitly _requires_ that the store be clonable.

Options are:

1. Lift that requirement up to the `impl` level.
2. Use `self.transient_slots.into_store()` to "take" the existing store.

Reply:
@Stebalien I've tried a few different variations of the "taking" option that you are suggesting. I didn't quite get anything to compile however.

Could you share a deeper explanation on the two options to either lift the requirement to the `impl` level (I'm not sure what you mean by this; which requirement? which `impl`?) and also if you can help me with the second option about taking the ownership of the datastore in order to satisfy the requirements of new_with_config. I'm still learning the ins and outs of this in rust, and would be very grateful. I would definitely not mind a code commit either if it's a simple lift for you. Thanks in advanced!

Reply:
Ah, so, this is going to be a _bit_ of a pain. The issue is that you can't leave `transient_slots` uninitialized so you need something to put there while you construct the new `Kamt`. E.g., you can have `Option<StateKamt<...>>` and put none there while you replace it.

But... doing that will also help facilitate lazy loading (you can leave it as `None` until the user first tries to load a transient slot). So it's not the end of the world.

Reply:
Thanks! I'm now returning from the thanksgiving break and will look into everything this week.

Reply:
I have learned many things since starting this :) I see that my previous attempt was very much against the RAII ideas that rust enforces. I didn't realize that was the issue. We don't need to destroy and create a new KAMT object but simply need to clear the one we have.

As a placeholder I have implemented a [loop over keys and delete implementation](https://github.com/filecoin-project/builtin-actors/pull/1588/files#diff-4c5e4db1ab6bf6de3f25e9c323a6d293b9acad0853f529377f7e54fafc986bfbR506) of `clear` for now until an efficient `clear()` can be used.  I opened a PR https://github.com/filecoin-project/ref-fvm/pull/2092 for an efficient implementation of `clear` in KAMT that I hope can be approved and it will be easy for me from there to upgrade kamt in builtin-actors to the future version that supports clear.